### PR TITLE
Add Kotlin -> Viper translation utilities

### PR DESCRIPTION
**Pull request by jesyspa**

### CODE REVIEW

File: plugins/formal-verification/src/org/jetbrains/kotlin/formver/scala/Collections.kt [hunk 1]
```diff
@@ -21,8 +21,8 @@ interface IntoScala<T> {
 }
 
 sealed class Option<in T> : IntoScala<scala.Option<@UnsafeVariance T>> {
-    data object None {
-        fun <T> toScala(): scala.Option<T> = scala.`None$`.`MODULE$` as scala.Option<T>
+    class None<T> : Option<T>() {
```
[the remark this answers is not in the document]
Comment (jesyspa) [reply plugins/formal-verification/src/org/jetbrains/kotlin/formver/scala/Collections.kt:24]: Ah, this was actually done prior to your changes to nullability, I just forgot to switch it back.
The reason then was that if a function took `Option<T>`, you could never pass `None` to it.
Now we don't actually pass around `Option<T>` so it shouldn't matter... not sure what's best.

File: plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+            Info.NoInfo.toViper(),
+            Trafos.NoTrafos.toViper()
+        )
+
+    fun addWithBody(declaration: FirSimpleFunction) {
+        val methodCtx = MethodConversionContext(this, declaration);
+        methods.add(methodCtx.fullMethod)
+    }
+
+    fun convertType(type: ConeKotlinType): ConvertedType {
```
Comment (gabryon99) [inline plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt]: I think you can rewrite this using a `when` expression. Like this:

```kotlin
fun convertType(type: ConeKotlinType): ConvertedType = when {
    type.isUnit -> ConvertedUnit()
    type.isInt -> ConvertedInt()
    else -> throw NotImplementedError()
}
```

File: plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        return null
+    }
+
+    override fun visitBlock(block: FirBlock, data: StmtConversionContext): Exp? {
+        // TODO: allow blocks to return values
+        block.statements.forEach { it.accept(this, data) }
+        return null
+    }
+
+    override fun <T> visitConstExpression(constExpression: FirConstExpression<T>, data: StmtConversionContext): Exp? {
```
Comment (gabryon99) [inline plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt:167]: Same thing for the `when` expression also here!

```kotlin
override fun <T> visitConstExpression(constExpression: FirConstExpression<T>, data: StmtConversionContext): Exp? 
 = when (constExpression.kind) {
    ConstantValueKind.Int -> Exp.IntLit((constExpression.value as Long).toInt().toScalaBigInt())
    else -> TODO("Not implemented yet")
}
```

File: plugins/formal-verification/src/org/jetbrains/kotlin/formver/scala/Collections.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/src/org/jetbrains/kotlin/formver/scala/Collections.kt:24]: Probably we should just remove it then

[the remark this answers is not in the document]
Comment (jesyspa) [reply plugins/formal-verification/src/org/jetbrains/kotlin/formver/scala/Collections.kt:24]: I'll note it down as a TODO, probably not blocking now

File: plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt (hunk 2 above)
Comment (jesyspa) [reply plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt]: Done

File: plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt (hunk 3 above)
Comment (jesyspa) [reply plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/ProgramConversionContext.kt]: Done

**Review by gabryon99 — APPROVED** [review]

**Pull request merged**